cmd/client/txgen/main.go

```diff
@@ -89,109 +97,103 @@ func main() {
+	utils.GetLogInstance().Debug("Cross Shard Ratio Is Set But not used", "cx ratio", *crossShardRatio)
 
 	// TODO(Richard): refactor this chuck to a single method
 	// Setup a logger to stdout and log file.
-	logFileName := fmt.Sprintf("./%v/txgen.log", *logFolder)
+	logFileName := fmt.Sprintf("./%v/txgen-%v.log", *logFolder, shardID)
 	h := log.MultiHandler(
 		log.StreamHandler(os.Stdout, log.TerminalFormat(false)),
 		log.Must.FileHandler(logFileName, log.LogfmtFormat()), // Log to file
 	)
 	log.Root().SetHandler(h)
 
 	gsif, err := consensus.NewGenesisStakeInfoFinder()
-	if err != nil {
-		_, _ = fmt.Fprintf(os.Stderr, "Cannot initialize stake info: %v\n", err)
-		os.Exit(1)
-	}
-
 	// Nodes containing blockchain data to mirror the shards' data in the network
-	nodes := []*node.Node{}
-	host, err := p2pimpl.NewHost(&selfPeer, nodePriKey)
+
+	myhost, err := p2pimpl.NewHost(&selfPeer, nodePriKey)
 	if err != nil {
 		panic("unable to new host in txgen")
 	}
-	for _, shardID := range shardIDs {
-		c := &consensus.Consensus{ShardID: shardID}
-		node := node.New(host, c, nil, false)
-		c.SetStakeInfoFinder(gsif)
-		c.ChainReader = node.Blockchain()
-		// Replace public keys with genesis accounts for the shard
-		c.PublicKeys = nil
-		startIdx := core.GenesisShardSize * shardID
-		endIdx := startIdx + core.GenesisShardSize
-		for _, acct := range contract.GenesisBLSAccounts[startIdx:endIdx] {
-			secretKey := bls2.SecretKey{}
-			if err := secretKey.SetHexString(acct.Private); err != nil {
-				_, _ = fmt.Fprintf(os.Stderr, "cannot parse secret key: %v\n",
-					err)
-				os.Exit(1)
-			}
-			c.PublicKeys = append(c.PublicKeys, secretKey.GetPublicKey())
-		}
-		// Assign many fake addresses so we have enough address to play with at first
-		nodes = append(nodes, node)
-	}
-
-	// Client/txgenerator server node setup
-	consensusObj, err := consensus.New(host, 0, p2p.Peer{}, nil)
 	if err != nil {
 		fmt.Fprintf(os.Stderr, "Error :%v \n", err)
 		os.Exit(1)
 	}
-
-	clientNode := node.New(host, consensusObj, nil, false)
-	clientNode.Client = client.NewClient(clientNode.GetHost(), shardIDs)
-
+	consensusObj, err := consensus.New(myhost, uint32(shardID), p2p.Peer{}, nil)
+	txGen := node.New(myhost, consensusObj, nil, false) //Changed it : no longer archival node.
+	txGen.Client = client.NewClient(txGen.GetHost(), shardIDs)
 	consensusObj.SetStakeInfoFinder(gsif)
-	consensusObj.ChainReader = clientNode.Blockchain()
-
+	consensusObj.ChainReader = txGen.Blockchain()
+	consensusObj.PublicKeys = nil
+	startIdx := 0
+	endIdx := startIdx + core.GenesisShardSize
+	for _, acct := range contract.GenesisBLSAccounts[startIdx:endIdx] {
+		secretKey := bls2.SecretKey{}
+		if err := secretKey.SetHexString(acct.Private); err != nil {
+			_, _ = fmt.Fprintf(os.Stderr, "cannot parse secret key: %v\n",
+				err)
+			os.Exit(1)
+		}
+		consensusObj.PublicKeys = append(consensusObj.PublicKeys, secretKey.GetPublicKey())
+	}
+	txGen.NodeConfig.SetRole(nodeconfig.ClientNode)
```

Comment:
Please put these bulk of code about txGen into a construction method.

Reply:
Done.